pkg/test/kubeharness.go

```diff
@@ -0,0 +1,272 @@
+		}
+
+		t.Cleanup(func() {
+			if err := env.Stop(); err != nil {
+				h.Errorf("error stopping envtest environment: %v", err)
+			}
+		})
+
+		h.restConfig = restConfig
+	} else if targetKube := os.Getenv("E2E_KUBE_TARGET"); targetKube == "mock" {
+		k8s, err := mockkubeapiserver.NewMockKubeAPIServer(":0")
+		if err != nil {
+			h.Fatalf("error building mock kube-apiserver: %v", err)
+		}
+
+		addr, err := k8s.StartServing()
+		if err != nil {
+			h.Fatalf("error starting mock kube-apiserver: %v", err)
+		}
+
+		t.Cleanup(func() {
+			if err := k8s.Stop(); err != nil {
+				h.Errorf("error stopping mock kube-apiserver: %v", err)
+			}
+		})
+
+		h.restConfig = &rest.Config{
+			Host: addr.String(),
+			ContentConfig: rest.ContentConfig{
+				ContentType: "application/json",
+			},
+			// gotta go fast during tests -- we don't really care about overwhelming our test API server
+			QPS:   1000.0,
+			Burst: 2000.0,
+		}
+	} else {
+		t.Fatalf("E2E_KUBE_TARGET=%q not supported", targetKube)
+	}
+
+	// Set up logging of k8s requests
+	logKubeRequests := true
+	if logKubeRequests {
+		eventSinks := EventSinksFromContext(ctx)
+		kubeEvents := NewMemoryEventSink()
+		h.KubeEvents = kubeEvents
+
+		eventSinks = append(eventSinks, kubeEvents)
+
+		wrapTransport := func(rt http.RoundTripper) http.RoundTripper {
+			t := NewHTTPRecorder(rt, eventSinks...)
+			return t
+		}
+		h.restConfig.Wrap(wrapTransport)
+	}
+
+	if h.client == nil {
+		client, err := client.New(h.restConfig, client.Options{})
+		if err != nil {
+			h.Fatalf("error building client: %v", err)
+		}
+		h.client = client
+	}
+
+	logging.SetupLogger()
+
+	if loadCRDs {
+		crds, err := crdloader.LoadAllCRDs()
+		if err != nil {
+			h.Fatalf("error loading crds: %v", err)
+		}
+		{
+			var wg sync.WaitGroup
+			var errsMutex sync.Mutex
+			var errs []error
+
+			for i := range crds {
+				crd := &crds[i]
+				wg.Add(1)
+				log.V(2).Info("loading crd", "name", crd.GetName())
+
+				go func() {
+					defer wg.Done()
+					if err := h.client.Create(ctx, crd.DeepCopy()); err != nil {
+						errsMutex.Lock()
+						defer errsMutex.Unlock()
+						errs = append(errs, fmt.Errorf("error creating crd %v: %w", crd.GroupVersionKind(), err))
+					}
+					h.waitForCRDReady(crd)
+				}()
+			}
+			wg.Wait()
+			if len(errs) != 0 {
+				h.Fatalf("error creating crds: %v", errors.Join(errs...))
+			}
+		}
+	}
+
+	return h
+}
+
+func (h *KubeHarness) GetClient() client.Client {
+	return h.client
+}
+
+func (h *KubeHarness) GetRESTConfig() *rest.Config {
+	return h.restConfig
+}
+
+func (h *KubeHarness) waitForCRDReady(obj client.Object) {
+	logger := log.FromContext(h.Ctx)
+
+	apiVersion, kind := obj.GetObjectKind().GroupVersionKind().ToAPIVersionAndKind()
+	name := obj.GetName()
+	namespace := obj.GetNamespace()
+
+	id := types.NamespacedName{Name: name, Namespace: namespace}
+	if err := wait.PollImmediate(2*time.Second, 1*time.Minute, func() (bool, error) {
```

Comment:
The `PollImmediate` is deprecated. The lib suggests to use `PollWithContextTimeout` instead.